### ISSUE TITLE
Dynamically changing multisampling causes a crash 

### ISSUE THREAD

**Issue by looeee** [pmndrs/react-postprocessing#97]
I tried to improve the performance scaling example by [reducing multisampling from 8 to 2 on regress](https://codesandbox.io/s/performance-scaling-reduce-multisampling-on-regress-6m1in). 

``` js
function Effects() {
  const [full, setFull] = useState(true)

  useFrame((state) => {
    // Reduce multisampling on regress
    state.performance.current < 1 ? setFull(false) : setFull(true)
  }, [])

  return (
    <EffectComposer multisampling={full ? 8 : 2}>
      <SSAO intensity={15} radius={10} luminanceInfluence={0} bias={0.035} />
      <Bloom kernelSize={KernelSize.LARGE} luminanceThreshold={0.55} luminanceSmoothing={0.2} />
    </EffectComposer>
  )
}
```

This works initially but after a minute or so of switching between the two values it crashes with `Cannot read property 'alpha' of null` (possibly the error is caused by recreating too many contexts):

``` 
VM1475:63 WARNING: Too many active WebGL contexts. Oldest context will be lost.
HTMLCanvasElement.getContext @ VM1475:63
postprocessing.esm.js:3089 Uncaught TypeError: Cannot read property 'alpha' of null
    at EffectComposer.createBuffer (postprocessing.esm.js:3089)
    at new EffectComposer (postprocessing.esm.js:3015)
    at eval (index.js:246)
    at updateMemo (react-reconciler.development.js:7294)
    at Object.useMemo (react-reconciler.development.js:7840)
    at useMemo (react.development.js:1532)
    at eval (index.js:245)
    at renderWithHooks (react-reconciler.development.js:6412)
    at updateForwardRef (react-reconciler.development.js:8471)
    at beginWork (react-reconciler.development.js:10525)
    at HTMLUnknownElement.callCallback (react-reconciler.development.js:12184)
    at Object.invokeGuardedCallbackDev (react-reconciler.development.js:12233)
    at invokeGuardedCallback (react-reconciler.development.js:12292)
    at beginWork$1 (react-reconciler.development.js:16531)
    at performUnitOfWork (react-reconciler.development.js:15337)
    at workLoopSync (react-reconciler.development.js:15268)
    at renderRootSync (react-reconciler.development.js:15231)
    at performSyncWorkOnRoot (react-reconciler.development.js:14840)
    at workLoop (scheduler.development.js:417)
    at flushWork (scheduler.development.js:390)
    at MessagePort.performWorkUntilDeadline (scheduler.development.js:157)
index.js:27 The above error occurred in the <ForwardRef> component:

    at eval (https://6m1in.csb.app/node_modules/@react-three/postprocessing/dist/index.js:242:212)
    at Effects (https://6m1in.csb.app/src/App.js:174:42)
    at Suspense
    at ErrorBoundary (https://6m1in.csb.app/node_modules/@react-three/fiber/dist/react-three-fiber.esm.js:1178:5)
    at Provider (https://6m1in.csb.app/node_modules/@react-three/fiber/dist/react-three-fiber.esm.js:1370:27)
 ```

**Comment by drcmda** [comment]
can it be this? https://github.com/pmndrs/react-postprocessing/blob/master/src/EffectComposer.tsx#L54

the isWebGL2Available creates a context, do the browsers only have a cap on canvas elements inside the dom or do they actually cap every invocation dom or otherwise? if that's the case a simple global let hasWebGL2 = undefined that's set once could maybe work. or can multisampling be set without having to recreate the effectcomposer? that would be the best option.

**Comment by looeee** [comment]
Looks like `isWebGL2Available` is the culprit. Easy to fix though:

![too many contexts](https://user-images.githubusercontent.com/5307958/131778438-892c22f5-5b89-44ed-92f5-3cd216829530.png)

``` js
const isWebGL2Available_V2 = () => {
  try {
    const canvas = document.createElement('canvas')
    const gl = canvas.getContext('webgl2');
    isAvailable = !!(window.WebGL2RenderingContext && gl);
    
    if (gl) gl.getExtension('WEBGL_lose_context').loseContext();
    
    return isAvailable;
  } catch (e) {
    return false
  }
}
```

**Comment by looeee** [comment]
Looking a bit deeper into the `EffectComposer` code, I guess my original idea of changing multisampling on the fly is not a great idea as it will cause the old render target to be disposed and a new one created every time. 
Similar to the suggestion for dynamically changing the passes, it would probably be required to manually create at least one of the buffers and reuse them. Something like this ([sandbox](https://codesandbox.io/s/performance-scaling-reduce-multisampling-on-regress-fixed-rx2hs)):

``` js
function Effects() {
  const ref = useRef()

  const [bufferInLow, bufferInHigh, bufferOutLow, bufferOutHigh] = useMemo(() => {
    if (!ref.current) return [null, null, null, null]
    const bufferInHigh = ref.current.inputBuffer.clone()
    const bufferOutHigh = ref.current.outputBuffer.clone()

    const bufferInLow = ref.current.createBuffer(bufferInHigh.depthBuffer, bufferInHigh.stencilBuffer, bufferInHigh.texture.type, 2)
    const bufferOutLow = ref.current.createBuffer(bufferOutHigh.depthBuffer, bufferOutHigh.stencilBuffer, bufferOutHigh.texture.type, 2)

  
    return [bufferInLow, bufferInHigh, bufferOutLow, bufferOutHigh]
  }, [ref.current])

  const [full, setFull] = useState(true)

  useFrame((state) => {
    state.performance.current < 1 ? setFull(false) : setFull(true)
  })

  useEffect(() => {
    if (!ref.current || !bufferInLow || !bufferInHigh || !bufferOutLow || !bufferOutHigh) return

    if (full) {
      ref.current.inputBuffer = bufferInHigh
      ref.current.outputBuffer = bufferOutHigh
    } else {
      ref.current.inputBuffer = bufferInLow
      ref.current.outputBuffer = bufferOutLow
    }
  }, [bufferInLow, bufferInHigh, bufferOutLow, bufferOutHigh, full])

  return (
    <EffectComposer ref={ref} multisampling={8}>
      <SSAO intensity={15} radius={10} luminanceInfluence={0} bias={0.035} />
      <Bloom kernelSize={KernelSize.LARGE} luminanceThreshold={0.55} luminanceSmoothing={0.2} />
    </EffectComposer>
  )
}
```

Although I'm getting intermittent errors due to undefined buffers so it probably needs a bit of work.

**Comment by looeee** [comment]
Anyway, we should probably still the isWebGL2Available method but in this case it was a canary that showed my code was bad!

**Comment by drcmda** [comment]
https://github.com/pmndrs/three-stdlib/commit/375e14084384fd6749d7219a7e71480f4e82fd98

**Comment by looeee** [comment]
Confirmed fixed in 2.1.2 👍
